File: muzei-api/src/main/java/com/google/android/apps/muzei/api/MuzeiContract.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.api;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Looper;
import android.provider.BaseColumns;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.FileNotFoundException;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresPermission;
import androidx.annotation.StringDef;
import androidx.annotation.WorkerThread;

/**
 * Contract between Muzei and applications, containing the definitions for all supported URIs and
 * columns as well as helper methods to make it easier to work with the provided data.
 */
public class MuzeiContract {
    /**
     * Only Muzei can write to the Muzei ContentProvider - this is a signature permission that no
     * other app can hold.
     */
    public static final String WRITE_PERMISSION = "com.google.android.apps.muzei.WRITE_PROVIDER";
    /**
     * Base authority for this content provider
     */
    public static final String AUTHORITY = "com.google.android.apps.muzei";
    /**
     * The scheme part for this provider's URI
     */
    private static final String SCHEME = "content://";

    /**
     * This class cannot be instantiated
     */
    private MuzeiContract() {
    }

    /**
     * Constants and helper methods for the Artwork table, providing access to the current artwork.
     *
     * <p>The Artwork table contains the details of the artwork that has been loaded by Muzei.
     * It also provides direct access to the cached image already downloaded by Muzei,
     * ensuring that you do not need to do additional networks requests or have internet access
     * when retrieving previously loaded artwork.
     *
     * <h3>Working with the Artwork table</h3>
     *
     * Querying {@link #CONTENT_URI} will return either zero rows (in cases where the user has never
     * activated Muzei before) or a row for each previously loaded artwork with all of the details needed to create an
     * {@link com.google.android.apps.muzei.api.Artwork Artwork} object. The helper method
     * {@link #getCurrentArtwork(Context)} does builds an Artwork object for the most recent artwork, although
     * you can certainly query the {@link #CONTENT_URI} directly and either use the columns directly or
     * use {@link com.google.android.apps.muzei.api.Artwork#fromCursor(android.database.Cursor) Artwork.fromCursor(Cursor)}
     * to parse the Cursor for you.
     *
     * <p>If instead you use {@link ContentResolver#openInputStream(Uri) ContentResolver.openInputStream(Uri)},
     * you'll get an InputStream for the cached image Bitmap. This can
     * then be passed to {@link BitmapFactory#decodeStream(java.io.InputStream) BitmapFactory.decodeStream(InputStream)} or
     * similar methods to retrieve the Bitmap itself. The helper method
     * {@link #getCurrentArtworkBitmap(Context)} does this operation, although note
     * that this may return a very large Bitmap so following the
     * <a href="http://developer.android.com/training/displaying-bitmaps/index.html">Displaying Bitmaps Efficiently training</a>
     * advice is highly suggested.
     *
     * <h3>Listening for changes</h3>
     *
     * Just like any {@link android.content.ContentProvider ContentProvider}, listening for changes can be done by
     * implementing a {@link android.database.ContentObserver ContentObserver} on {@link #CONTENT_URI} or by using
     * any of the helper classes such as {@link android.content.CursorLoader CursorLoader} to listen for updates.
     *
     * <p>However, if you want to receive updates while in the background and do not want to maintain
     * a constantly running {@link android.database.ContentObserver ContentObserver}, you can instead listen for the
     * {@link #ACTION_ARTWORK_CHANGED} broadcast, sent out immediately after an update is made:
     *
     * <pre class="prettyprint">
     * &lt;receiver android:name=".ExampleArtworkUpdateReceiver"&gt;
     *     &lt;intent-filter&gt;
     *         &lt;action android:name="com.google.android.apps.muzei.ACTION_ARTWORK_CHANGED" /&gt;
     *     &lt;/intent-filter&gt;
     * &lt;/receiver&gt;
     * </pre>
     *
     * No data is sent alongside the broadcast, but this can be used to kick off an
     * {@link android.app.IntentService IntentService} to retrieve the latest artwork or start
     * other processing.
     */
    public static final class Artwork implements BaseColumns {
        /**
         * The set of valid font types to use to display artwork meta info.
         *
         * @see #COLUMN_NAME_META_FONT
         * @see #META_FONT_TYPE_DEFAULT
         * @see #META_FONT_TYPE_ELEGANT
         */
        @Retention(RetentionPolicy.SOURCE)
        @StringDef({META_FONT_TYPE_DEFAULT, META_FONT_TYPE_ELEGANT})
        public @interface MetaFontType {}
        /**
         * The default font type for {@link #COLUMN_NAME_META_FONT}
         */
        public static final String META_FONT_TYPE_DEFAULT = "";
        /**
         * An elegant alternate font type for {@link #COLUMN_NAME_META_FONT}
         */
        public static final String META_FONT_TYPE_ELEGANT = "elegant";
        /**
         * Column name for the authority of the provider for this artwork.
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_SOURCE_COMPONENT_NAME = "sourceComponentName";
        /**
         * Column name of the artwork image URI. In almost all cases you should use
         * {@link ContentResolver#openInputStream(Uri) ContentResolver.openInputStream(CONTENT_URI)}
         * to retrieve the already downloaded artwork.
         * <p>Type: TEXT (URI)
         */
        public static final String COLUMN_NAME_IMAGE_URI = "imageUri";
        /**
         * Column name for the artwork's title.
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_TITLE = "title";
        /**
         * Column name for the artwork's byline (such as author and date).
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_BYLINE = "byline";
        /**
         * Column name for the artwork's attribution info.
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_ATTRIBUTION = "attribution";
        /**
         * Column name for the artwork's opaque application-specific identifier.
         * This is generally only useful to the app that published the artwork and should
         * not be relied upon by other apps.
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_TOKEN = "token";
        /**
         * Column name for the artwork's view Intent, encoded via
         * {@link android.content.Intent#toUri(int) Intent.toUri(Intent.URI_INTENT_SCHEME)} and can
         * be decoded via
         * {@link android.content.Intent#parseUri(String, int) Intent.parseUri(viewIntent, Intent.URI_INTENT_SCHEME)}
         * <p>Type: TEXT (Intent encoded URI)
         */
        public static final String COLUMN_NAME_VIEW_INTENT = "viewIntent";
        /**
         * Column name for the font type to use to display artwork meta info.
         * <p>Type: TEXT (one of {@link #META_FONT_TYPE_DEFAULT} or {@link #META_FONT_TYPE_ELEGANT})
         */
        public static final String COLUMN_NAME_META_FONT = "metaFont";
        /**
         * Column name for when this artwork was added.
         * This will be automatically added for you by Muzei.
         * <p>Type: LONG (in milliseconds)
         */
        public static final String COLUMN_NAME_DATE_ADDED = "date_added";
        /**
         * The MIME type of {@link #CONTENT_URI} providing artwork.
         */
        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.google.android.apps.muzei.artwork";
        /**
         * The MIME type of {@link #CONTENT_URI} providing a single piece of artwork.
         */
        public static final String CONTENT_ITEM_TYPE = "vnd.android.cursor.item/vnd.google.android.apps.muzei.artwork";
        /**
         * The default sort order for this table
         */
        public static final String DEFAULT_SORT_ORDER = COLUMN_NAME_DATE_ADDED + " DESC";
        /**
         * The table name offered by this provider.
         */
        public static final String TABLE_NAME = "artwork";

        /**
         * This class cannot be instantiated
         */
        private Artwork() {
        }

        /**
         * The content:// style URL for this table. This is the main entry point for queries and for
         * opening an {@link java.io.InputStream InputStream} to the current artwork's image.
         * <p>
         * Apps can only {@link ContentResolver#query query} for artwork; only Muzei can write to the ContentProvider.
         *
         * @see #getCurrentArtwork
         * @see #getCurrentArtworkBitmap
         */
        @RequiresPermission.Write(@RequiresPermission(WRITE_PERMISSION))
        public static final Uri CONTENT_URI = Uri.parse(MuzeiContract.SCHEME + MuzeiContract.AUTHORITY
                + "/" + Artwork.TABLE_NAME);
        /**
         * Intent action that will be broadcast when the artwork is changed. This happens immediately after the
         * ContentProvider is updated with data and should be considered the signal that you can retrieve the new
         * artwork.
         */
        public static final String ACTION_ARTWORK_CHANGED = "com.google.android.apps.muzei.ACTION_ARTWORK_CHANGED";

        /**
         * Returns the current Muzei {@link com.google.android.apps.muzei.api.Artwork Artwork}.
         *
         * @param context the context to retrieve a ContentResolver.
         *
         * @return the current {@link com.google.android.apps.muzei.api.Artwork Artwork}
         * or null if one could not be found.
         */
        public static com.google.android.apps.muzei.api.Artwork getCurrentArtwork(Context context) {
            ContentResolver contentResolver = context.getContentResolver();
            Cursor cursor = contentResolver.query(CONTENT_URI, null, null, null, null);
            try {
                if (cursor == null || !cursor.moveToFirst()) {
                    return null;
                }
                return com.google.android.apps.muzei.api.Artwork.fromCursor(cursor);
            } finally {
                if (cursor != null) {
                    cursor.close();
                }
            }
        }

        /**
         * Naively gets the current artwork image without any subsampling or optimization for output size
         * <p>Note that this may return a very large Bitmap so following the
         * <a href="http://developer.android.com/training/displaying-bitmaps/index.html">Displaying Bitmaps Efficiently training</a>
         * is highly recommended.
         *
         * @param context the context to retrieve a ContentResolver.
         *
         * @return A Bitmap of the current artwork or null if the image could not be decoded.
         *
         * @throws FileNotFoundException If no cached artwork image was found.
         */
        @WorkerThread
        public static Bitmap getCurrentArtworkBitmap(Context context) throws FileNotFoundException {
            if (Looper.myLooper() == Looper.getMainLooper()) {
                throw new IllegalStateException("getCurrentArtworkBitmap cannot be called on the main thread");
            }
            ContentResolver contentResolver = context.getContentResolver();
            return BitmapFactory.decodeStream(contentResolver.openInputStream(CONTENT_URI));
        }
    }

    /**
     * Constants and helper methods for the Sources table, providing access to sources' information.
     */
    public static final class Sources implements BaseColumns {
        /**
         * Column name for the authority of the provider.
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_COMPONENT_NAME = "component_name";
        /**
         * Column name for the flag indicating if the source is currently selected
         * <p>Type: INTEGER (boolean)
         */
        public static final String COLUMN_NAME_IS_SELECTED = "selected";
        /**
         * Column name for the source's description.
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_DESCRIPTION = "description";
        /**
         * Column name for the flag indicating if the source wants callbacks for network connectivity changes
         * <p>Type: INTEGER (boolean)
         */
        public static final String COLUMN_NAME_WANTS_NETWORK_AVAILABLE = "network";
        /**
         * Column name for the flag indicating if the source supports a 'Next Artwork' action
         * <p>Type: INTEGER (boolean)
         */
        public static final String COLUMN_NAME_SUPPORTS_NEXT_ARTWORK_COMMAND = "supports_next_artwork";
        /**
         * Column name for the commands the source supports
         * <p>Type: TEXT
         */
        public static final String COLUMN_NAME_COMMANDS = "commands";
        /**
         * The MIME type of {@link #CONTENT_URI} providing sources.
         */
        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.google.android.apps.muzei.source";
        /**
         * The MIME type of {@link #CONTENT_URI} providing a single source.
         */
        public static final String CONTENT_ITEM_TYPE = "vnd.android.cursor.item/vnd.google.android.apps.muzei.source";
        /**
         * The default sort order for this table
         */
        public static final String DEFAULT_SORT_ORDER = Sources.COLUMN_NAME_IS_SELECTED + " DESC," +
                Sources.COLUMN_NAME_COMPONENT_NAME;
        /**
         * The table name offered by this provider.
         */
        public static final String TABLE_NAME = "sources";

        /**
         * This class cannot be instantiated
         */
        private Sources() {
        }

        /**
         * The content:// style URL for this table.
         * <p>
         * Apps can only {@link ContentResolver#query query} for source info; only Muzei can write to the
         * ContentProvider.
         */
        @RequiresPermission.Write(@RequiresPermission(WRITE_PERMISSION))
        public static final Uri CONTENT_URI = Uri.parse(MuzeiContract.SCHEME + MuzeiContract.AUTHORITY
                + "/" + Sources.TABLE_NAME);
        /**
         * Intent action that will be broadcast when the source info is changed. This happens immediately after the
         * ContentProvider is updated with data and should be considered the signal that you can retrieve the new
         * source info.
         */
        public static final String ACTION_SOURCE_CHANGED = "com.google.android.apps.muzei.ACTION_SOURCE_CHANGED";

        /**
         * Parse the commands found in the {@link #COLUMN_NAME_COMMANDS} field into a List of {@link UserCommand}s.
         *
         * @param commandsString The serialized commands found in {@link #COLUMN_NAME_COMMANDS}.
         *
         * @return A deserialized List of {@link UserCommand}s.
         */
        @NonNull
        public static List<UserCommand> parseCommands(String commandsString) {
            ArrayList<UserCommand> commands = new ArrayList<>();
            if (commandsString == null) {
                return commands;
            }
            try {
                JSONArray commandArray = new JSONArray(commandsString);
                for (int h=0; h<commandArray.length(); h++) {
                    commands.add(UserCommand.deserialize(commandArray.getString(h)));
                }
            } catch (JSONException e) {
                Log.e(MuzeiContract.Sources.class.getSimpleName(), "Error parsing commands from " + commandsString, e);
            }
            return commands;
        }
    }
}

File: muzei-api/src/main/java/com/google/android/apps/muzei/api/MuzeiContract.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.api

import android.content.ContentProvider
import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.database.ContentObserver
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Looper
import android.provider.BaseColumns
import androidx.annotation.RequiresPermission
import androidx.annotation.WorkerThread
import com.google.android.apps.muzei.api.MuzeiContract.Artwork.ACTION_ARTWORK_CHANGED
import com.google.android.apps.muzei.api.MuzeiContract.Artwork.CONTENT_URI
import com.google.android.apps.muzei.api.MuzeiContract.Artwork.getCurrentArtwork
import com.google.android.apps.muzei.api.MuzeiContract.Artwork.getCurrentArtworkBitmap
import com.google.android.apps.muzei.api.provider.ProviderClient
import java.io.FileNotFoundException

/**
 * Contract between Muzei and applications, containing the definitions for all supported URIs and
 * columns as well as helper methods to make it easier to work with the provided data.
 */
public object MuzeiContract {
    /**
     * Only Muzei can write to the Muzei ContentProvider - this is a signature permission that no
     * other app can hold.
     */
    public const val WRITE_PERMISSION: String = "com.google.android.apps.muzei.WRITE_PROVIDER"
    /**
     * Base authority for this content provider
     */
    public const val AUTHORITY: String = "com.google.android.apps.muzei"
    /**
     * The scheme part for this provider's URI
     */
    private const val SCHEME = "content://"

    /**
     * Constants and helper methods for the Artwork table, providing access to the current artwork.
     *
     * The Artwork table contains the details of the artwork that has been loaded by Muzei.
     * It also provides direct access to the cached image already downloaded by Muzei,
     * ensuring that you do not need to do additional networks requests or have internet access
     * when retrieving previously loaded artwork.
     *
     * ### Working with the Artwork table
     *
     * Querying [CONTENT_URI] will return either zero rows (in cases where the user has never
     * activated Muzei before) or a row for each previously loaded artwork with all of the
     * details needed to create an [Artwork][com.google.android.apps.muzei.api.Artwork] object.
     * The helper method [getCurrentArtwork] builds an Artwork object for the most recent
     * artwork, although you can certainly query the [CONTENT_URI] directly and either
     * use the columns directly or use
     * [Artwork.fromCursor][com.google.android.apps.muzei.api.Artwork.fromCursor]
     * to parse the Cursor for you.
     *
     * If instead you use [ContentResolver.openInputStream], you'll get an InputStream for the
     * cached image Bitmap. This can then be passed to [BitmapFactory.decodeStream] or
     * similar methods to retrieve the Bitmap itself. The helper method [getCurrentArtworkBitmap]
     * does this operation, although note that this may return a very large Bitmap so following the
     * [Handling Bitmaps documentation](https://developer.android.com/topic/performance/graphics)
     * advice is highly suggested.
     *
     * ### Listening for changes
     *
     * Just like any [ContentProvider], listening for changes can be done by implementing a
     * [ContentObserver] on [CONTENT_URI] to listen for updates.
     *
     * On API 24+ devices, it is strongly recommended to use `WorkManager` or
     * `JobScheduler` to listen for content URI changes in the background without
     * maintaining a constantly running [ContentObserver].
     *
     * To support earlier versions of Android, you can listen for the
     * [ACTION_ARTWORK_CHANGED] broadcast, sent out immediately after an update is made:
     *
     * ```
     * <receiver android:name=".ExampleArtworkUpdateReceiver">
     *   <intent-filter>
     *     <action android:name="com.google.android.apps.muzei.ACTION_ARTWORK_CHANGED" />
     *   </intent-filter>
     * </receiver>
     * ```
     *
     * No data is sent alongside the broadcast, but this can be used to kick off
     * background processing to retrieve the latest artwork or start other processing.
     */
    public object Artwork {
        /**
         * The unique ID of the Artwork, suitable for use with [ContentUris.withAppendedId].
         */
        @Suppress("unused", "ObjectPropertyName")
        public const val _ID: String = BaseColumns._ID
        /**
         * Column name for the authority of the provider for this artwork.
         *
         * Type: TEXT
         */
        public const val COLUMN_NAME_PROVIDER_AUTHORITY: String = "sourceComponentName"
        /**
         * Column name of the artwork image URI. In almost all cases you should use
         * [ContentResolver.openInputStream(CONTENT_URI)][ContentResolver.openInputStream]
         * to retrieve the already downloaded artwork.
         *
         * Type: TEXT (URI)
         */
        public const val COLUMN_NAME_IMAGE_URI: String = "imageUri"
        /**
         * Column name for the artwork's title.
         *
         * Type: TEXT
         */
        public const val COLUMN_NAME_TITLE: String = "title"
        /**
         * Column name for the artwork's byline (such as author and date).
         *
         * Type: TEXT
         */
        public const val COLUMN_NAME_BYLINE: String = "byline"
        /**
         * Column name for the artwork's attribution info.
         *
         * Type: TEXT
         */
        public const val COLUMN_NAME_ATTRIBUTION: String = "attribution"
        /**
         * Column name for when this artwork was added.
         *
         * Type: LONG (in milliseconds)
         */
        public const val COLUMN_NAME_DATE_ADDED: String = "date_added"
        /**
         * The MIME type of [CONTENT_URI] providing artwork.
         */
        public const val CONTENT_TYPE: String = "vnd.android.cursor.dir/vnd.google.android.apps.muzei.artwork"
        /**
         * The MIME type of [CONTENT_URI] providing a single piece of artwork.
         */
        public const val CONTENT_ITEM_TYPE: String = "vnd.android.cursor.item/vnd.google.android.apps.muzei.artwork"
        /**
         * The default sort order for this table
         */
        @Suppress("unused")
        public const val DEFAULT_SORT_ORDER: String = "$COLUMN_NAME_DATE_ADDED DESC"
        /**
         * The table name offered by this provider.
         */
        public const val TABLE_NAME: String = "artwork"

        /**
         * The content:// style URL for this table. This is the main entry point for queries and for
         * opening an [InputStream][java.io.InputStream] to the current artwork's image.
         *
         * Apps can only [query][ContentResolver.query] for artwork; only Muzei can write
         * to the ContentProvider.
         *
         * @see getCurrentArtwork
         * @see getCurrentArtworkBitmap
         */
        @JvmStatic
        @get:JvmName("getContentUri")
        @RequiresPermission.Write(RequiresPermission(WRITE_PERMISSION))
        public val CONTENT_URI: Uri = Uri.parse("$SCHEME$AUTHORITY/$TABLE_NAME")
        /**
         * Intent action that will be broadcast when the artwork is changed. This happens
         * immediately after the ContentProvider is updated with data and should be considered
         * the signal that you can retrieve the new artwork.
         */
        @Deprecated("This broadcast cannot be received on Build.VERSION_CODES.O " +
                "and higher devices. Use WorkManager or JobScheduler to listen for " +
                "artwork change events in the background on API 24+ devices.")
        public const val ACTION_ARTWORK_CHANGED: String = "com.google.android.apps.muzei.ACTION_ARTWORK_CHANGED"

        /**
         * Returns the current Muzei [Artwork][com.google.android.apps.muzei.api.Artwork].
         *
         * @param context the context to retrieve a ContentResolver.
         *
         * @return the current [Artwork][com.google.android.apps.muzei.api.Artwork]
         * or null if one could not be found.
         */
        @JvmStatic
        public fun getCurrentArtwork(context: Context): com.google.android.apps.muzei.api.Artwork? {
            val contentResolver = context.contentResolver
            return contentResolver.query(CONTENT_URI, null, null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    com.google.android.apps.muzei.api.Artwork.fromCursor(cursor)
                } else {
                    null
                }
            }
        }

        /**
         * Naively gets the current artwork image without any subsampling or optimization for output size
         *
         * Note that this may return a very large Bitmap so following the
         * [Displaying Bitmaps Efficiently training](http://developer.android.com/training/displaying-bitmaps/index.html)
         * is highly recommended.
         *
         * @param context the context to retrieve a ContentResolver.
         *
         * @return A Bitmap of the current artwork or null if the image could not be decoded.
         *
         * @throws FileNotFoundException If no cached artwork image was found.
         */
        @JvmStatic
        @WorkerThread
        @Throws(FileNotFoundException::class)
        public fun getCurrentArtworkBitmap(context: Context): Bitmap? {
            check(Looper.myLooper() != Looper.getMainLooper()) {
                "getCurrentArtworkBitmap cannot be called on the main thread"
            }
            val contentResolver = context.contentResolver
            return BitmapFactory.decodeStream(contentResolver.openInputStream(CONTENT_URI))
        }
    }

    /**
     * Constants and helper methods for the Sources table, providing access to sources' information.
     */
    public object Sources {
        /**
         * The unique ID of the Source, suitable for use with [ContentUris.withAppendedId].
         */
        @Suppress("unused", "ObjectPropertyName")
        public const val _ID: String = BaseColumns._ID
        /**
         * Column name for the authority of the provider for this artwork.
         *
         * Type: TEXT
         */
        public const val COLUMN_NAME_AUTHORITY: String = "component_name"
        /**
         * Column name for the source's description.
         *
         * Type: TEXT
         */
        public const val COLUMN_NAME_DESCRIPTION: String = "description"
        /**
         * Column name for the flag indicating if the source supports a 'Next Artwork' action
         *
         * Type: INTEGER (boolean)
         */
        public const val COLUMN_NAME_SUPPORTS_NEXT_ARTWORK_COMMAND: String = "supports_next_artwork"
        /**
         * The MIME type of [CONTENT_URI] providing sources.
         */
        public const val CONTENT_TYPE: String = "vnd.android.cursor.dir/vnd.google.android.apps.muzei.source"
        /**
         * The MIME type of [CONTENT_URI] providing a single source.
         */
        public const val CONTENT_ITEM_TYPE: String = "vnd.android.cursor.item/vnd.google.android.apps.muzei.source"
        /**
         * The default sort order for this table
         */
        @Suppress("unused")
        public const val DEFAULT_SORT_ORDER: String = COLUMN_NAME_AUTHORITY
        /**
         * The table name offered by this provider.
         */
        public const val TABLE_NAME: String = "sources"

        /**
         * The content:// style URL for this table.
         *
         * Apps can only [query][ContentResolver.query] for source info; only Muzei can write
         * to the ContentProvider.
         */
        @JvmStatic
        @get:JvmName("getContentUri")
        @RequiresPermission.Write(RequiresPermission(WRITE_PERMISSION))
        public val CONTENT_URI: Uri = Uri.parse("$SCHEME$AUTHORITY/$TABLE_NAME")
        /**
         * Intent action that will be broadcast when the source info is changed. This happens
         * immediately after the ContentProvider is updated with data and should be considered
         * the signal that you can retrieve the new source info.
         */
        @Deprecated("This broadcast cannot be received on Build.VERSION_CODES.O " +
                "and higher devices. Use WorkManager or JobScheduler to listen for " +
                "source change events in the background on API 24+ devices.")
        public const val ACTION_SOURCE_CHANGED: String = "com.google.android.apps.muzei.ACTION_SOURCE_CHANGED"

        /**
         * Checks the [Sources] table provided by Muzei to determine whether the
         * [com.google.android.apps.muzei.api.provider.MuzeiArtProvider] associated with the
         * given [authority] has been selected by the user.
         *
         * If Muzei is not installed or Muzei has never been activated, this will return
         * `false`.
         */
        @JvmStatic
        public fun isProviderSelected(context: Context, authority: String): Boolean = context.contentResolver.query(
                CONTENT_URI, arrayOf(COLUMN_NAME_AUTHORITY),
                "$COLUMN_NAME_AUTHORITY=?", arrayOf(authority),
                null)?.use { data ->
            val authorityColumn = data.getColumnIndex(COLUMN_NAME_AUTHORITY)
            while (data.moveToNext()) {
                val selectedAuthority = data.getString(authorityColumn)
                if (selectedAuthority == authority) {
                    return true
                }
            }
            false
        } ?: false

        /**
         * Deep link into Muzei's Sources screen, automatically scrolling
         * to the [com.google.android.apps.muzei.api.provider.MuzeiArtProvider] associated
         * with the given [authority]. If Muzei is not yet activated, users will be asked
         * to activate Muzei before continuing onto the Sources screen.
         *
         * Note that users must still manually select your `MuzeiArtProvider`.
         *
         * It is strongly recommended in cases where you receive an
         * [android.content.ActivityNotFoundException] (due to the user having an older version
         * of Muzei installed or if Muzei is not installed at all) to fall back on:
         *
         * 1. [android.content.pm.PackageManager.getLaunchIntentForPackage] with Muzei's
         * package name, `net.nurik.roman.muzei` to open Muzei and navigate to the
         * Sources screen themselves.
         * 2. A link to the Google Play Store with Muzei's package name so that user can
         * install Muzei.
         */
        @JvmStatic
        public fun createChooseProviderIntent(authority: String): Intent = Intent(Intent.ACTION_VIEW)
                .setData(Uri.parse(BuildConfig.CHOOSE_PROVIDER_URI_PREFIX + authority))
    }
}

/**
 * Checks the [MuzeiContract.Sources] table provided by Muzei to determine whether the
 * [com.google.android.apps.muzei.api.provider.MuzeiArtProvider] associated with
 * this [ProviderClient] has been selected by the user.
 *
 * If Muzei is not installed or Muzei has never been activated, this will return
 * `false`.
 */
@Suppress("unused")
public fun ProviderClient.isSelected(context: Context): Boolean =
        MuzeiContract.Sources.isProviderSelected(context, contentUri.authority!!)

/**
 * Deep link into Muzei's Sources screen, automatically scrolling
 * to the [com.google.android.apps.muzei.api.provider.MuzeiArtProvider] associated
 * with this [ProviderClient]. If Muzei is not yet activated, users will be asked
 * to activate Muzei before continuing onto the Sources screen.
 *
 * Note that users must still manually select your `MuzeiArtProvider`.
 *
 * It is strongly recommended in cases where you receive an
 * [android.content.ActivityNotFoundException] (due to the user having an older version
 * of Muzei installed or if Muzei is not installed at all) to fall back on:
 *
 * 1. [android.content.pm.PackageManager.getLaunchIntentForPackage] with Muzei's
 * package name, `net.nurik.roman.muzei` to open Muzei and navigate to the
 * Sources screen themselves.
 * 2. A link to the Google Play Store with Muzei's package name so that user can
 * install Muzei.
 */
@Suppress("unused")
public fun ProviderClient.createChooseProviderIntent(): Intent =
        MuzeiContract.Sources.createChooseProviderIntent(contentUri.authority!!)
